src/lib/Simplify.hs

```diff
@@ -708,97 +705,17 @@ hasExceptions expr = do
-  (IxCache fv m) <> (IxCache fv' m') = IxCache (fv <> fv') (m <> m')
-instance Monoid (IxCache n) where
-  mempty = IxCache mempty mempty
-instance SubstE Name IxCache where
-  -- This is needed, because IxCache is sometimes below decls that are emitted in-place.
-  -- But all those decls should hopefully already be fresh, since we try to be good about
-  -- generating fresh names before emitting them, so the slow branch of refreshAbs should
-  -- not be taken. If it is, don't implement this method, but better treat this as a
-  -- performance bug instead!
-  substE _ _ = error "needed for static constraints, hopefully dynamically unreachable!"
-
-type MonadIxCache1 (m::MonadKind1) = forall n. MonadState (IxCache n) (m n)
-
-instance Monad1 m => HoistableState IxCache m where
-  -- TODO: I think we can do hoisting only based on the free vars in keys.
-  -- Instances should only be added at the top-level so it's not like they
-  -- can refer to any local vars that could prevent the values from being hoistable.
-  hoistState s b s' = case hoist b s' of
-    HoistSuccess s'' -> return s''
-    HoistFailure _   -> return s
-
-simplifiedIxInstance
-  :: (EnvReader m, MonadIxCache1 m)
-  => Type n -> m n (SimpleIxInstance n)
-simplifiedIxInstance ty = do
-  let key = EKey ty
-  gets (HM.lookup key . ixCacheMap) >>= \case
-    Just a  -> return a
-    Nothing -> {-# SCC simplifyInstance #-} do
-      a <- liftSimplifyM simplifyInstance
-      modify (<> IxCache (freeVarsE key <> freeVarsE a) (HM.singleton key a))
-      return a
-  where
-    simplifyInstance = liftSimplifyM do
-      Abs decls inst <- liftBuilder $ buildScoped $ getIxImpl $ sink ty
-      simpAbs <- buildScoped $ simplifyDecls decls do
-        (s , IdentityReconAbs) <- simplifyLam $ ixSize inst
-        (to, IdentityReconAbs) <- simplifyLam $ toOrdinal inst
-        (fo, IdentityReconAbs) <- simplifyLam $ unsafeFromOrdinal inst
-        return $! IxImpl{ ixSize = s, toOrdinal = to, unsafeFromOrdinal = fo }
-      return $! case simpAbs of
-        Abs simpDecls IxImpl{..} ->
-          SimpleIxInstance
-            (Abs simpDecls $ fromLam ixSize           )
-            (Abs simpDecls $ fromLam toOrdinal        )
-            (Abs simpDecls $ fromLam unsafeFromOrdinal)
-
-    fromLam = \case Lam l -> l; _ -> error "Not a lambda!"
-{-# SCC simplifiedIxInstance #-}
-
-appSimplifiedIxMethod
-  :: (Emits n, Builder m, MonadIxCache1 m)
-  => Type n -> (SimpleIxInstance n -> Abs (Nest Decl) LamExpr n)
-  -> Atom n -> m n (Atom n)
-appSimplifiedIxMethod ty method x = do
-  Abs decls f <- method <$> simplifiedIxInstance ty
-  f' <- emitDecls decls f
-  Distinct <- getDistinct
-  case f' of
-    LamExpr fx' fb' -> emitBlock =<< applySubst (fx' @> SubstVal x) fb'
+emitSimplified
+  :: (Emits n, Builder m)
+  => (forall l. (Emits l, DExt n l) => BuilderM l (Atom l))
+  -> m n (Atom n)
+emitSimplified m = emitBlock . dceApproxBlock =<< buildBlockSimplified m
+{-# INLINE emitSimplified #-}
```

Comment:
You're right that there is a performance hit, but it seems to be fairly small. I measured it at 5% on the kernel regression example. As you say, there's no longer a cache for simplification, so we're doing more simplification than we used to. But on the other hand, we're doing less Ix synthesis than we used to. Imagine a function using `Fin n` that gets inlined at multiple places. Previously, each of those use sites had to do their own Ix synthesis (no cache hits because `n` is local in each place). Now we just do Ix synthesis once. I had actually hoped that it would come out as a net win, but for now it doesn't seem to be. I'll try to think about ways to get the best of both.